### PR TITLE
[flutter_tools] support sound null-safety mode for the web

### LINKED ISSUE #59873: default web null-safety to --sound-null-safety and skip autodetection
Right now, the tool needs to configure the paths to the dill used by the frontend server for compilation, but doesn't know whether the app is going to be compiled in sound or unsound mode. The most straightforward solution would be to provide both dills and let the frontend_server decide.

EDIT:

With the vm using an agnostic dill, there should be no issues with null-safety detection. The web on the other hand requires separate dills and separate precompiled SDKs. In this case we should remove support for auto-detection and default to --sound-null-safety when the experiment is enabled. This will require the tool to choose the correct dill and serve the correct dart_sdk.js

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

In web debug mode, infer sound null safety by default. When sound null safety is enabled, provide a separate dill and precompiled Dart SDK. Release builds do not need this setting since we run dart2js from source.

Fixes https://github.com/flutter/flutter/issues/59873

### CODE REVIEW

File: packages/flutter_tools/lib/src/artifacts.dart [hunk 1]
```diff
@@ -42,6 +42,8 @@ enum Artifact {
   flutterWebLibrariesJson,
   /// The summary dill for the dartdevc target.
   webPlatformKernelDill,
+  /// The summdary dill with null safety enabled for the dartdevc target.
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/artifacts.dart:45]:
```suggestion
  /// The summary dill with null safety enabled for the dartdevc target.
```

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart [hunk 2]
```diff
@@ -466,6 +469,14 @@ class WebAssetServer implements AssetReader {
     'dart_sdk.js',
   ));
 
+  @visibleForTesting
+  final File dartSdkSound = globals.fs.file(globals.fs.path.join(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_runner/devfs_web.dart:473]: can we make this its own artifact?

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:473]: Yeah, that is a good idea

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart [hunk 3]
```diff
@@ -474,6 +485,14 @@ class WebAssetServer implements AssetReader {
     'dart_sdk.js',
   ));
 
+  @visibleForTesting
+  final File canvasKitDartSdkSound = globals.fs.file(globals.fs.path.join(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/build_runner/devfs_web.dart:489]: same here

File: packages/flutter_tools/lib/src/build_runner/devfs_web.dart (hunk 2 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/build_runner/devfs_web.dart:473]: Thanks, it'll help me with my refactor

**Review by christopherfujino — APPROVED** [review]
LGTM other than nit about creating individual artifacts for files we're referencing.

**Comment by jonahwilliams** [comment]
Added artifacts for all of the precompiled SDKs

**Comment by jonahwilliams** [comment]
@renyou not sure what is up with the google testing check. It seems to have only patched in part of the diff. One of the errors is:

```
Error: Final field 'nullSafetyMode' is not initialized.
Try to initialize the field in the declaration or in every constructor.
  final NullSafetyMode nullSafetyMode;
```

And checking the CL shows that the constructor initialization is left out.

**Pull request merged**